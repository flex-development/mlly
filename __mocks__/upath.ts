/**
 * @file Mocks - upath
 * @module mocks/upath
 * @see https://github.com/anodynos/upath
 */

/**
 * [`upath`][1] module type.
 *
 * [1]: https://github.com/anodynos/upath
 */
type Actual = typeof import('upath')

/**
 * `upath` module.
 *
 * @const {Actual} actual
 */
const actual: Actual = await vi.importActual<Actual>('upath')

export default {
  basename: vi.fn(actual.basename),
  dirname: vi.fn(actual.dirname),
  extname: vi.fn(actual.extname),
  format: vi.fn(actual.format),
  isAbsolute: vi.fn(actual.isAbsolute),
  join: vi.fn(actual.join),
  parse: vi.fn(actual.parse),
  relative: vi.fn(actual.relative),
  resolve: vi.fn(actual.resolve)
}
