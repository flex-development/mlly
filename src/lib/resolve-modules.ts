/**
 * @file resolveModules
 * @module mlly/lib/resolveModules
 */

import type { ResolveOptions as Options } from '#src/interfaces'
import type { SpecifierType } from '#src/types'
import upath from 'upath'
import extractStatements from './extract-statements'
import resolveModule from './resolve-module'

/**
 * Resolves all modules in `code`.
 *
 * @see [`resolveModule`](./resolve-module.ts)
 *
 * @async
 *
 * @param {string} code - Code containing module specifiers
 * @param {Omit<Options, 'type'>} [options={}] - Resolve options
 * @return {Promise<string>} `code` with modules resolved
 */
const resolveModules = async (
  code: string,
  options: Options = {}
): Promise<string> => {
  for (const statement of extractStatements(code)) {
    if (!statement.specifier) continue

    /**
     * Determines a module specifier type for `resolved`.
     *
     * @param {string} resolved - Resolved module
     * @return {SpecifierType} Module specifier type for `resolved`
     */
    const type = (resolved: string): SpecifierType => {
      return upath.isAbsolute(statement.specifier!) ||
        statement.specifier!.startsWith('file:')
        ? 'absolute'
        : /\/node_modules\//.test(resolved)
        ? 'bare'
        : 'relative'
    }

    code = code.replace(
      statement.code,
      statement.code.replace(
        statement.specifier,
        await resolveModule(statement.specifier, {
          ...options,
          type: options.type ?? type
        })
      )
    )
  }

  return code
}

export default resolveModules
