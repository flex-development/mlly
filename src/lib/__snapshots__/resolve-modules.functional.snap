// Vitest Snapshot v1

exports[`functional:lib/resolveModules > should resolve modules in code 1`] = `
"import isBuiltin from '@flex-development/is-builtin'
import { moduleResolve, type ErrnoException } from 'import-meta-resolve'
import { pathToFileURL, URL } from 'node:url'
import upath from 'upath'
import { CONDITIONS, RESOLVE_EXTENSIONS } from '../constants.mjs'
import type { ResolveOptions as Options } from '../interfaces/index.mjs'
import type { SpecifierType } from '../types/index.mjs'
import toBareSpecifier from './to-bare-specifier.mjs'
import toRelativeSpecifier from './to-relative-specifier.mjs'"
`;
